python/src/py_nvbench.cpp

```diff
@@ -0,0 +1,723 @@
+  py_benchmark_cls.def("get_name", &nvbench::benchmark_base::get_name);
+  py_benchmark_cls.def(
+    "add_int64_axis",
+    [](nvbench::benchmark_base &self, std::string name, std::vector<nvbench::int64_t> data) {
+      self.add_int64_axis(std::move(name), std::move(data));
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("name"),
+    py::arg("values"));
+  py_benchmark_cls.def(
+    "add_int64_power_of_two_axis",
+    [](nvbench::benchmark_base &self, std::string name, std::vector<nvbench::int64_t> data) {
+      self.add_int64_axis(std::move(name),
+                          std::move(data),
+                          nvbench::int64_axis_flags::power_of_two);
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("name"),
+    py::arg("values"));
+  py_benchmark_cls.def(
+    "add_float64_axis",
+    [](nvbench::benchmark_base &self, std::string name, std::vector<nvbench::float64_t> data) {
+      self.add_float64_axis(std::move(name), std::move(data));
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("name"),
+    py::arg("values"));
+  py_benchmark_cls.def(
+    "add_string_axis",
+    [](nvbench::benchmark_base &self, std::string name, std::vector<std::string> data) {
+      self.add_string_axis(std::move(name), std::move(data));
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("name"),
+    py::arg("values"));
+  py_benchmark_cls.def(
+    "set_name",
+    [](nvbench::benchmark_base &self, std::string name) {
+      self.set_name(std::move(name));
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("name"));
+  py_benchmark_cls.def(
+    "set_is_cpu_only",
+    [](nvbench::benchmark_base &self, bool is_cpu_only) {
+      self.set_is_cpu_only(is_cpu_only);
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("is_cpu_only"));
+  // TODO: should this be exposed?
+  py_benchmark_cls.def(
+    "set_run_once",
+    [](nvbench::benchmark_base &self, bool run_once) {
+      self.set_run_once(run_once);
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("run_once"));
+  py_benchmark_cls.def(
+    "set_skip_time",
+    [](nvbench::benchmark_base &self, nvbench::float64_t skip_duration_seconds) {
+      self.set_skip_time(skip_duration_seconds);
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("duration_seconds"));
+  py_benchmark_cls.def(
+    "set_timeout",
+    [](nvbench::benchmark_base &self, nvbench::float64_t duration_seconds) {
+      self.set_timeout(duration_seconds);
+      return std::ref(self);
+    },
+    py::return_value_policy::reference,
+    py::arg("duration_seconds"));
+  py_benchmark_cls.def(
+    "set_throttle_threshold",
+    [](nvbench::benchmark_base &self, nvbench::float64_t threshold) {
```

Comment:
question: both `set_throttle_threshold` and `set_throttle_recovery_delay` accept `nvbench::float32_t` but here we have `nvbench::float64_t`. Is this intentional?

Reply:
I've changed signature to `float32_t`. 

With `nvbench::float64_t` Python extension would cast `object` of type `float` to `nvbench::float64_t`, which always succeeds since underlying type is `double` in both cases. Then, during the call to the C++ method compiler would silently cast `nvbench::float64_t` to `nvbench::float32_t` possibly flushing values outside of range representable in `float32_t` to zero or to infinity.

After the change, Python extension would check value of input value of `object` type, and would raise `ValueError` if the Python argument can not be represented by `nvbench::float32_t`.